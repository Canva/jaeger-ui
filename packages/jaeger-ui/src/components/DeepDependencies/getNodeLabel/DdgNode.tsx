// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';
import cx from 'classnames';

import focalNodeIcon from './focalNodeIcon';
import setFocusButtonIcon from './setFocusButtonIcon';
import NewWindowIcon from '../../common/NewWindowIcon';
import { getUrl as getSearchUrl } from '../../SearchTracePage/url';
import { EViewModifier, PathElem } from '../../../model/ddg/types';

import './DdgNode.css';

type TProps = {
  vertexKey: string;
  service: string;
  operation: string | null;
  focalNodeUrl: string | null;
  getVisiblePathElems: (vertexKey: string) => PathElem[] | undefined;
  isFocalNode: boolean;
  isUiFindMatch: boolean;
  viewModifiers: number;
  setViewModifier: (vertexKey: string, viewModifier: EViewModifier, isEnabled: boolean) => void;
};

// While browsers suport URLs of unlimited length, many server clients do not handle more than this max
const MAX_LENGTH = 2083;
const MIN_LENGTH = getSearchUrl().length;
const PARAM_NAME_LENGTH = '&traceID='.length;

export default class DdgNode extends React.PureComponent<TProps> {
  viewTraces = () => {
    const { vertexKey, getVisiblePathElems } = this.props;
    const elems = getVisiblePathElems(vertexKey);
    if (elems) {
      const ids: Set<string> = new Set();
      let currLength = MIN_LENGTH;
      for (let i = 0; i < elems.length; i++) {
        const id = elems[i].memberOf.traceID;
        // Keep track of the length, then break if it is too long, to avoid opening a tab with a URL that the
        // backend cannot process, even if there are more traceIDs
        currLength += PARAM_NAME_LENGTH + id.length;
        if (currLength > MAX_LENGTH) break;
        ids.add(id);
      }
      window.open(getSearchUrl({ traceID: Array.from(ids) }), '_blank');
    }
  };

  onMouseUx = (event: React.MouseEvent<HTMLElement>) => {
    const { vertexKey, setViewModifier } = this.props;
    setViewModifier(vertexKey, EViewModifier.Hovered, event.type === 'mouseover');
  };

  render() {
    const { focalNodeUrl, isFocalNode, isUiFindMatch, operation, service } = this.props;
    return (
      <div className="DdgNode" onMouseOver={this.onMouseUx} onMouseOut={this.onMouseUx}>
        <div
          className={cx('DdgNode--core', { 'is-focalNode': isFocalNode, 'is-UiFindMatch': isUiFindMatch })}
        >
          {isFocalNode && <div className="DdgNode--focalMarker">{focalNodeIcon}</div>}
          <div>
            <h4 className="DdgNode--label">{service}</h4>
            {operation && <div className="DdgNode--label">{operation}</div>}
          </div>
        </div>

        <div className="DdgNode--actionsWrapper">
          {focalNodeUrl && (
            <a href={focalNodeUrl} className="DdgNode--actionsItem">
              {setFocusButtonIcon}
              <span className="DdgNode--actionsItemText">Set focus</span>
            </a>
          )}
          <a className="DdgNode--actionsItem" onClick={this.viewTraces}>
            <NewWindowIcon />
            <span className="DdgNode--actionsItemText">View traces</span>
          </a>
        </div>
      </div>
    );
  }
}
