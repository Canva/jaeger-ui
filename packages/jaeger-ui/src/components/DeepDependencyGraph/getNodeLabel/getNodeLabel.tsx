// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';

import DdgNode from './DdgNode';
import { DdgVertex, StyleStates } from '../../../model/ddg/types';

// temp fill in props
const noops = {
  setAsFocalNode(service: string, operation: string | null) {
    // eslint-disable-next-line no-console
    console.log(`set as focal: ${service} -- ${operation}`);
  },
  setViewModifier(id: string, viewModifier: StyleStates, enabled: boolean) {
    // eslint-disable-next-line no-console
    console.log(`set view modifier: ${enabled ? 'on' : 'OFF'} ${viewModifier} -- ${id}`);
  },
};

// temp -- transform DdgVertex for rendering purposes
export default function getNodeLabel(vertex: DdgVertex) {
  const pathElem = [...vertex.pathElems.values()][0];
  if (!pathElem) {
    throw new Error('Invalid vertex state');
  }
  const { name: operationName, service } = pathElem.operation;
  const isFocalNode = pathElem.memberIdx === pathElem.memberOf.focalIdx;
  return (
    <DdgNode
      id={vertex.key}
      service={service.name}
      operation={operationName}
      isFocalNode={isFocalNode}
      viewModifiers={0}
      {...noops}
    />
  );
}
