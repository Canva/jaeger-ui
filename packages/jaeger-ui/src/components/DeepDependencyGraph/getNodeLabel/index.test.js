// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { shallow } from 'enzyme';

import getNodeLabel from './index';

describe('getNodeLabel()', () => {
  it('returns a <DdgNode/>', () => {
    const opName = 'the-operation';
    const svcName = 'the-service';
    const focalIdx = 3;
    const pathElem = {
      memberOf: { focalIdx },
      memberIdx: 2,
      operation: {
        name: opName,
        service: {
          name: svcName,
        },
      },
    };
    const pathElems = new Set([pathElem]);
    let ddgNode = getNodeLabel({ pathElems });
    expect(ddgNode).toBeDefined();
    let wrapper = shallow(ddgNode);
    expect(wrapper).toMatchSnapshot();
    // set as the focal node
    pathElem.memberIdx = focalIdx;
    ddgNode = getNodeLabel({ pathElems });
    expect(ddgNode).toBeDefined();
    wrapper = shallow(ddgNode);
    expect(wrapper).toMatchSnapshot();
  });

  it('throws an error if given a vertex without any path elements', () => {
    const pathElems = new Set();
    expect(() => getNodeLabel({ pathElems })).toThrow();
  });
});
