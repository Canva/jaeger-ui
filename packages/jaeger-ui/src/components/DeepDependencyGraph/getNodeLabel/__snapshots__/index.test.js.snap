// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getNodeLabel() returns a <DdgNode/> 1`] = `
<div
  className="DdgNode"
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <div
    className="DdgNode--core"
  >
    <div>
      <h4
        className="DdgNode--label"
      >
        the-service
      </h4>
      <div
        className="DdgNode--label"
      >
        the-operation
      </div>
    </div>
  </div>
  <div
    className="DdgNode--actionsWrapper"
  >
    <a
      className="DdgNode--actionsItem"
      href="some-url"
    >
      <svg
        className="DdgSetFocusButtonIcon"
        height="100"
        viewBox="0 0 100 100"
        width="100"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g>
          <path
            d="M50.0001 -1L61.0557 22.1383H38.9444L50.0001 -1Z"
            fill="currentColor"
          />
          <path
            d="M49.9999 99L38.9443 75.8617L61.0556 75.8617L49.9999 99Z"
            fill="currentColor"
          />
          <path
            d="M100 49L76.8617 60.0556L76.8617 37.9444L100 49Z"
            fill="currentColor"
          />
          <path
            d="M1.57361e-06 49L23.1383 37.9444L23.1383 60.0556L1.57361e-06 49Z"
            fill="currentColor"
          />
        </g>
      </svg>
      <span
        className="DdgNode--actionsItemText"
      >
        Set focus
      </span>
    </a>
    <a
      className="DdgNode--actionsItem"
    >
      <NewWindowIcon
        isLarge={false}
      />
      <span
        className="DdgNode--actionsItemText"
      >
        View traces
      </span>
    </a>
  </div>
</div>
`;
